pyvista/core/filters/image_data.py

```diff
@@ -2923,3 +2923,513 @@ def _validate_dimensional_operation(
+                if processing_cell_scalars:
+                    dimensions_ = dimensions_ - 1 if processing_cell_scalars else dimensions_
+                reference_image.dimensions = dimensions_  # type: ignore[assignment]
+            new_dimensions = np.array(reference_image.dimensions)
+
+        # Compute the magnification factors to use with vtkImageResample
+        # Note that SetMagnificationFactors will multiply the factors by the extent
+        # but we want to multiply the dimensions. These values are off by one.
+        singleton_dims = old_dimensions == 1
+        magnification_factors = (new_dimensions - 1) / (old_dimensions - 1)
+        magnification_factors[singleton_dims] = 1
+
+        resample = _vtk.vtkImageResample()
+        resample.SetInputData(input_image)
+        resample.SetMagnificationFactors(*magnification_factors)
+
+        # Set interpolation mode
+        if interpolation == 'linear':
+            resample.SetInterpolationModeToLinear()
+        elif interpolation == 'cubic':
+            resample.SetInterpolationModeToCubic()
+        else:  # 'nearest'
+            resample.SetInterpolationModeToNearestNeighbor()
+
+        # Get output
+        _update_alg(resample, progress_bar=progress_bar)
+        output_image = _get_output(resample).copy(deep=False)
+
+        # Set geometry from the reference
+        output_image.direction_matrix = reference_image.direction_matrix
+        output_image.origin = reference_image.origin
+        output_image.offset = reference_image.offset
+
+        if reference_image_provided:
+            output_image.spacing = reference_image.spacing
+        else:
+            # Need to fixup the spacing
+            old_spacing = np.array(input_image.spacing)
+            output_dimensions = np.array(output_image.dimensions)
+
+            if extend_border and not processing_cell_scalars:
+                # Compute spacing to have the same effective sample rate as the dimensions
+                actual_sample_rate = output_dimensions / old_dimensions
+                new_spacing = old_spacing / actual_sample_rate
+
+                # This will enlarge the image, so we need to shift the origin accordingly
+                # Shift the origin by 1/2 of the old and new spacing, but keep the spacing
+                # unchanged for singleton dimensions.
+                shift_old = old_spacing[~singleton_dims] / 2
+                shift_new = new_spacing[~singleton_dims] / 2
+                new_origin = np.array(input_image.origin)
+                new_origin[~singleton_dims] += shift_new - shift_old
+
+                output_image.origin = new_origin
+            else:
+                # Compute spacing to match bounds of input and dimensions of output
+                bnds = input_image.bounds
+                size = np.array(
+                    (bnds.x_max - bnds.x_min, bnds.y_max - bnds.y_min, bnds.z_max - bnds.z_min)
+                )
```

Comment:
```suggestion
                size = np.array(input_image.size)
```
Use size property from #7092

Reply:
This change can also be made later as part of https://github.com/pyvista/pyvista/pull/7092 in case we want to merge this PR before that one